crates/virtual-node/src/lib.rs

```diff
@@ -187,79 +239,124 @@ impl VirtualNode {
 
-                element.append_child(&child_elem).unwrap();
+                    element.append_child(&child_elem).unwrap();
+                },
             }
         });
 
-        CreatedElement { element, closures }
+        CreatedNode { node: element, closures }
     }
 
+}
+
+impl VirtualNodeText {
     /// Return a `Text` element from a `VirtualNode`, typically right before adding it
     /// into the DOM.
     pub fn create_text_node(&self) -> Text {
         let document = web_sys::window().unwrap().document().unwrap();
-        document.create_text_node(&self.text.as_ref().unwrap())
+        document.create_text_node(&self.text)
     }
+}
 
-    /// Whether or not this `VirtualNode` is representing a `Text` node
-    pub fn is_text_node(&self) -> bool {
-        self.text.is_some()
+/// A node along with all of the closures that were created for that
+/// node's events and all of it's child node's events.
+pub struct CreatedNode<T> {
+    /// A `Node` or `Element` that was created from a `VirtualNode`
+    pub node: T,
+    /// A map of a node's unique identifier along with all of the Closures for that node.
+    ///
+    /// The DomUpdater uses this to look up nodes and see if they're still in the page. If not
+    /// the reference that we maintain to their closure will be dropped, thus freeing the Closure's
+    /// memory.
+    pub closures: HashMap<u32, Vec<DynClosure>>,
+}
+
+impl<T> CreatedNode<T> {
+    pub fn without_closures<N: Into<T>>(node: N) -> Self {
+        CreatedNode {
+            node: node.into(),
+            closures: HashMap::with_capacity(0),
+        }
     }
+}
 
-    /// Whether or not this is a self closing tag such as <br> or <img />
-    pub fn is_self_closing(&self) -> bool {
-        SELF_CLOSING_TAGS.contains(self.tag.as_str())
+impl<T> Deref for CreatedNode<T> {
+    type Target = T;
+    fn deref(&self) -> &Self::Target {
+        &self.node
     }
 }
 
-impl From<&str> for VirtualNode {
-    fn from(text: &str) -> Self {
-        VirtualNode::text(text)
+impl From<CreatedNode<Element>> for CreatedNode<Node> {
+    fn from(other: CreatedNode<Element>) -> CreatedNode<Node> {
+        CreatedNode {
+            node: other.node.into(),
+            closures: other.closures,
+        }
     }
 }
-impl From<String> for VirtualNode {
-    fn from(text: String) -> Self {
-        VirtualNode::text(&text)
+
+fn create_unique_identifier() -> u32 {
+    let mut elem_unique_id = ELEM_UNIQUE_ID.lock().unwrap();
+
+    *elem_unique_id += 1;
+
+    *elem_unique_id
+}
+
+impl From<VirtualNodeText> for VirtualNode {
+    fn from(other: VirtualNodeText) -> Self {
+        VirtualNode::Text(other)
     }
 }
-impl<'a> From<&'a String> for VirtualNode {
-    fn from(text: &'a String) -> Self {
-        VirtualNode::text(text)
+
+impl From<VirtualNodeElement> for VirtualNode {
+    fn from(other: VirtualNodeElement) -> Self {
+        VirtualNode::Element(other)
+    }
+}
+
+impl From<&str> for VirtualNodeText {
```

Comment:
Would `From<AsRef<str>>` be useful here?

Reply:
I just tested it. It [does work](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d061965b074cf5fe0b87682d0e1e63bd) but has the disadvantage that if you pass in a `String` by value, it will be copied and then dropped.

In contrast, with the explicit `From<String>` impl we can move the string into the `VText`.

I'd stick to the two separate impls even though it's a bit more code.

Reply:
Oh sorry I meant in addition, not as a replacement

Reply:
Ah, that won't work unfortunately, since Rust complains about conflicting impls (since str and String both implement AsRef).